src/games.c

```diff
@@ -1252,35 +1266,31 @@ int igraph_degree_sequence_game_no_multiple_undirected_uniform(igraph_t *graph,
         IGRAPH_CHECK(igraph_set_init(set, 0));
         VECTOR(adjlist)[i] = set;
         IGRAPH_CHECK(igraph_set_reserve(set, (long) VECTOR(*degseq)[i]));
-    }
-    IGRAPH_FINALLY(igraph_vector_ptr_destroy_all, &adjlist);
+    }    
 
     RNG_BEGIN();
 
     for (;;) {
         igraph_bool_t success = 1;
-        IGRAPH_CHECK(igraph_vector_int_shuffle(&stubs));
 
-        /* optimization: we do an initial pass looking for self-loops */
+        /* Shuffle stubs vector with Fisher-Yates and check for self-loops and multi-edges as we go. */
```

Comment:
Why not simply use [`igraph_vector_shuffle`](https://igraph.org/c/doc/igraph-Data-structures.html#igraph_vector_shuffle)?

Reply:
I think that's the whole point: instead of shuffling the vector completely and _then_ checking for self-loops at the end, we check for self-loops as the shuffle goes. So, if a self-loop is produced early in the shuffle in a part of the vector that is not going to be touched any more, we can bail out earlier and start from scratch. Is that right @szhorvat?

Reply:
Ah, sorry, was too quick looking at it! :+1:

Reply:
Yes, that's the whole point, and that's what makes the speedup possible.